controllers/openstackcluster_controller.go

```diff
@@ -438,44 +438,69 @@ func reconcileNetworkComponents(log logr.Logger, osProviderClient *gophercloud.P
-		var port int32
-		if openStackCluster.Spec.APIServerPort == 0 {
-			port = 6443
-		} else {
-			port = int32(openStackCluster.Spec.APIServerPort)
-		}
-		fp, err := networkingService.GetOrCreateFloatingIP(openStackCluster, clusterName, openStackCluster.Spec.APIServerFloatingIP)
-		if err != nil {
-			handleUpdateOSCError(openStackCluster, errors.Errorf("Floating IP cannot be got or created: %v", err))
-			return errors.Errorf("Floating IP cannot be got or created: %v", err)
-		}
-		// Set APIEndpoints so the Cluster API Cluster Controller can pull them
-		openStackCluster.Spec.ControlPlaneEndpoint = clusterv1.APIEndpoint{
-			Host: fp.FloatingIP,
-			Port: port,
-		}
-	}
 
 	err = networkingService.ReconcileSecurityGroups(openStackCluster, clusterName)
 	if err != nil {
 		handleUpdateOSCError(openStackCluster, errors.Errorf("failed to reconcile security groups: %v", err))
 		return errors.Errorf("failed to reconcile security groups: %v", err)
 	}
 
+	// Calculate the port that we will use for the API server
+	var apiServerPort int
+	if openStackCluster.Spec.ControlPlaneEndpoint.IsValid() {
+		apiServerPort = int(openStackCluster.Spec.ControlPlaneEndpoint.Port)
+	} else if openStackCluster.Spec.APIServerPort != 0 {
+		apiServerPort = openStackCluster.Spec.APIServerPort
+	} else {
+		apiServerPort = 6443
+	}
+
 	if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
 		loadBalancerService, err := loadbalancer.NewService(osProviderClient, clientOpts, log)
 		if err != nil {
 			return err
 		}
 
-		err = loadBalancerService.ReconcileLoadBalancer(openStackCluster, clusterName)
+		err = loadBalancerService.ReconcileLoadBalancer(openStackCluster, clusterName, apiServerPort)
 		if err != nil {
 			handleUpdateOSCError(openStackCluster, errors.Errorf("failed to reconcile load balancer: %v", err))
 			return errors.Errorf("failed to reconcile load balancer: %v", err)
 		}
 	}
 
+	if !openStackCluster.Spec.ControlPlaneEndpoint.IsValid() {
+		var host string
+		// If there is a load balancer use the floating IP for it if set, falling back to the internal IP
+		if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
+			if openStackCluster.Status.Network.APIServerLoadBalancer.IP != "" {
+				host = openStackCluster.Status.Network.APIServerLoadBalancer.IP
+			} else {
+				host = openStackCluster.Status.Network.APIServerLoadBalancer.InternalIP
+			}
+		} else if !openStackCluster.Spec.DisableAPIServerFloatingIP {
+			// If floating IPs are not disabled, get one to use as the VIP for the control plane
+			fp, err := networkingService.GetOrCreateFloatingIP(openStackCluster, clusterName, openStackCluster.Spec.APIServerFloatingIP)
+			if err != nil {
+				handleUpdateOSCError(openStackCluster, errors.Errorf("Floating IP cannot be got or created: %v", err))
+				return errors.Errorf("Floating IP cannot be got or created: %v", err)
+			}
+			host = fp.FloatingIP
+		} else {
+			// This case is not managed for now (i.e. no load balancer + no floating IP)
```

Comment:
Can we skip this error? That's exactly what we're doing: using kube-vip and setting allowedAddressPairs. This is all handled at the template level: https://paste.opendev.org/raw/809295/

Reply:
@petrutlucian94

In the case you are describing, the test if !openStackCluster.Spec.ControlPlaneEndpoint.IsValid() will fail and you will not enter this block. If you get to the point where that error is returned, it actually is an error that will prevent the cluster from deploying.

In the case where you are using kube-vip, you will actually need the ability to specify the fixed IP ahead of time, which this PR doesn't support but could easily be added.

Reply:
Oh, right. This is under the ``!openStackCluster.Spec.ControlPlaneEndpoint.IsValid()`` branch, all good then. Sorry, got distracted by the comment. May be worth updating the comment though.

Reply:
@petrutlucian94

Having said that, as discussed before I would prefer to support an `APIServerFixedIP` option rather than requiring `ControlPlaneEndpoint.Host` to contain an IP address. In the case where there is no load balancer, not specifying it would be an error, so this check does need an additional clause.

Reply:
Yep, sounds good to me.